src/ShallowWaterRosenbrock.jl

```diff
@@ -0,0 +1,252 @@
+function shallow_water_rosenbrock_time_step!(
+     y₂, ϕ, F, q₁, q₂, duh₁, duh₂, H1h, H1hchol, y_wrk,  # in/out args
+     model, dΩ, dω, Y, V, Q, R, S, f, g, y₁, yₘ,         # in args
+     RTMMchol, L2MMchol, Amat, Bchol, Blfchol,           # more in args
+     dt, τ, leap_frog)                                   # ...yet more in args
+  # energetically balanced second order rosenbrock shallow water solver
+  # reference: eqns (24) and (39) of
+  # https://github.com/BOM-Monash-Collaborations/articles/blob/main/energetically_balanced_time_integration/EnergeticallyBalancedTimeIntegration_SW.tex
+  #
+  # f          : coriolis force (field)
+  # g          : gravity (constant)
+  # h₁         : fluid depth at current time level
+  # u₁         : fluid velocity at current time level
+  # RTMM       : H(div) mass matrix, ∫β⋅βdΩ, ∀β∈ H(div,Ω)
+  # L2MM       : L² mass matrix, ∫γγdΩ, ∀γ∈ L²(Ω)
+  # dt         : time step
+  # τ          : potential vorticity upwinding parameter
+  # dΩ         : measure of the elements
+
+  n = get_normal_vector(model)
+  dt₁ = dt
+  if leap_frog
+    dt₁ = 2.0*dt
+  end
+
+  # multifield terms
+  u₁, h₁ = y₁
+  u₂, h₂ = y₂
+  uₘ, hₘ = yₘ
+
+  du₁,dh₁ = duh₁
+  du₂,dh₂ = duh₂
+
+  # 1.1: the mass flux
+  compute_mass_flux!(F,dΩ,V,RTMMchol,u₁*h₁)
+  # 1.2: the bernoulli function
+  compute_bernoulli_potential!(ϕ,dΩ,Q,L2MMchol,u₁⋅u₁,h₁,g)
+  # 1.3: the potential vorticity
+  compute_potential_vorticity!(q₁,H1h,H1hchol,dΩ,R,S,h₁,u₁,f,n)
+  # 1.4: assemble the velocity residual
+  bᵤ₁(v) = ∫(-1.0*(q₁ - τ*u₁⋅∇(q₁))*(v⋅⟂(F,n)))dΩ + ∫(DIV(v)*ϕ)dω
+  # 1.5: assemble the depth residual
+  bₕ₁(q) = ∫(-q*DIV(F))dω
+
+  bₕᵤ₁((v,q)) = bᵤ₁(v) + bₕ₁(q)
+  Gridap.FESpaces.assemble_vector!(bₕᵤ₁, get_free_dof_values(duh₁), Y)
+
+  # Solve for du₁, dh₁ over a MultiFieldFESpace
+  ldiv!(Blfchol, get_free_dof_values(duh₁))
+
+  # update
+  get_free_dof_values(u₂) .= get_free_dof_values(uₘ) .+ dt₁ .* get_free_dof_values(du₁)
+  get_free_dof_values(h₂) .= get_free_dof_values(hₘ) .+ dt₁ .* get_free_dof_values(dh₁)
+
+  # 2.1: the mass flux
+  compute_mass_flux!(F,dΩ,V,RTMMchol,u₁*(2.0*h₁ + h₂)/6.0+u₂*(h₁ + 2.0*h₂)/6.0)
+  # 2.2: the bernoulli function
+  compute_bernoulli_potential!(ϕ,dΩ,Q,L2MMchol,(u₁⋅u₁ + u₁⋅u₂ + u₂⋅u₂)/3.0,0.5*(h₁ + h₂),g)
+  # 2.3: the potential vorticity
+  compute_potential_vorticity!(q₂,H1h,H1hchol,dΩ,R,S,h₂,u₂,f,n)
+  # 2.4: assemble the velocity residual
```

Comment:
Transform lines 61-68 by a call to the function mentioned in https://github.com/gridapapps/GridapGeosciences.jl/pull/18/files#r711858624